File: part-02/src/test/java/eu/luminis/workshop/smallsteps/logic/domainmodel/LegoPartsTest.java
package eu.luminis.workshop.smallsteps.logic.domainmodel;

import org.junit.jupiter.api.Test;

import java.util.LinkedHashMap;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.*;

class LegoPartsTest {

    @Test
    public void test_hasParts() {
        assertTrue(new LegoParts(Map.of("3306", 2)).hasParts());
        assertFalse(new LegoParts(Map.of()).hasParts());
    }

    @Test
    public void test_allPartsPresentInOtherLegoParts() {
        LegoParts other = new LegoParts(Map.of("3306", 2, "101708", 3, "783101", 4));

        assertTrue(new LegoParts(Map.of("3306", 2, "101708", 2)).allPartsPresentIn(other));
        assertTrue(new LegoParts(Map.of("101708", 3, "3306", 2)).allPartsPresentIn(other));
        assertFalse(new LegoParts(Map.of("3306", 2, "999999", 2)).allPartsPresentIn(other));
        assertTrue(new LegoParts(Map.of()).allPartsPresentIn(other));
        assertFalse(new LegoParts(Map.of("3306", 2, "999999", 2))
                .allPartsPresentIn(new LegoParts(Map.of())));
    }

    @Test
    public void test_hasNonePartWithZeroNumber() {
        assertTrue(new LegoParts(Map.of("3306", 2, "101708", 3)).nonePartsWithZero());
        assertFalse(new LegoParts(Map.of("3306", 0, "101708", 3)).nonePartsWithZero());
    }

    @Test
    public void test_allPartsAreAvailableInRequestedNumber() {
        LegoParts other = new LegoParts(Map.of("3306", 2, "101708", 3));

        assertTrue(other.allPartsInRequestedNumberAvailable(
                new LegoParts(Map.of("3306", 2, "101708", 3))));
        assertFalse(other.allPartsInRequestedNumberAvailable(
                new LegoParts(Map.of("3306", 2, "101708", 4))));
    }

    @Test
    public void test_add_two_LegoParts() {
        LegoParts one = new LegoParts(Map.of("3306", 2, "101708", 3));
        LegoParts two = new LegoParts(Map.of("3306", 2, "101716", 3));
        assertThat(one.plus(two).listParts()).containsExactlyInAnyOrderEntriesOf(Map.of(
                "3306", 4,
                "101708", 3,
                "101716", 3
        ));
    }

    @Test
    public void test_list_parts_sorted() {
        LegoParts one = new LegoParts(Map.of(
                "3306", 2,
                "101708", 3,
                "7800", 1,
                "1100", 9
                ));

        assertThat(one.listPartsSorted()).containsExactlyEntriesOf(
                new LinkedHashMap<>() {{
                    put("1100", 9);
                    put("3306", 2);
                    put("7800", 1);
                    put("101708", 3);
                }});
    }

    @Test
    public void test_equal() {
        assertEquals(
                new LegoParts(Map.of("3306", 2, "101708", 3)),
                new LegoParts(Map.of("3306", 2, "101708", 3))
        );
    }
}

File: part-02/src/test/kotlin/eu/luminis/workshop/smallsteps/logic/domainModel/LegoPartsTest.kt
package eu.luminis.workshop.smallsteps.logic.domainModel

import kotlin.test.Test
import kotlin.test.assertFalse
import kotlin.test.assertTrue

internal class LegoPartsTest {
    @Test
    fun isNotEmpty() {
        assertFalse(LegoParts().isNotEmpty())
        assertTrue(LegoParts(mapOf("23894" to 3)).isNotEmpty())
    }

    @Test
    fun hasNoneWithCountZero() {
        assertTrue(LegoParts().hasNoneWithCountZero())
        assertTrue(LegoParts(mapOf("1244" to 1)).hasNoneWithCountZero())
        assertFalse(LegoParts(mapOf("1244" to 0)).hasNoneWithCountZero())
        assertFalse(LegoParts(mapOf("1244" to 0, "2355" to 5)).hasNoneWithCountZero())
        assertTrue(LegoParts(mapOf("1244" to 8, "2355" to 5)).hasNoneWithCountZero())
    }

    @Test
    fun partsPresentIn() {
        assertTrue(LegoParts().partsPresentIn(LegoParts()))
        assertTrue(LegoParts(mapOf("1234" to 1)).partsPresentIn(LegoParts(mapOf("1234" to 1))))
        assertTrue(LegoParts(mapOf("1234" to 1)).partsPresentIn(LegoParts(mapOf("1234" to 1, "8757" to 2))))
        assertTrue(LegoParts(mapOf("1234" to 5, "8757" to 1)).partsPresentIn(LegoParts(mapOf("1234" to 0, "8757" to 0))))
        assertFalse(LegoParts(mapOf("1234" to 5, "8757" to 1)).partsPresentIn(LegoParts()))
        assertFalse(LegoParts(mapOf("1234" to 1)).partsPresentIn(LegoParts(mapOf("4321" to 1, "8757" to 2))))
    }

    @Test
    fun hasNoMoreThan() {
    }

    @Test
    fun plus() {
    }
}